runtime/tests/checker/reference_test.go

```diff
@@ -3804,3 +3804,37 @@ func TestCheckReferenceRequiredTypeAnnotation(t *testing.T) {
 		)
 	})
 }
+
+func TestCheckOptionalReference(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nested optional reference", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+            fun main() {
+                var dict: {String: Foo?} = {}
+                var ref: (&Foo)?? = &dict["foo"] as &Foo??
+            }
+
+            struct Foo {}
+        `)
+
+		require.NoError(t, err)
+	})
+
+	t.Run("reference to nested optional", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+            fun main() {
+                var dict: {String: Foo?} = {}
+                var ref: &(Foo??) = &dict["foo"] as &(Foo??)
```

Comment:
Did we not previously ban references to optionals, and change them to be optional references? IMO the same logic should apply here; taking a reference to a `T??` type should produce a `(&T)??` type.

Reply:
> Did we not previously ban references to optionals, and change them to be optional references?

It was supposed to, but this path wasn't handled. This is actually the second part of the issue: https://github.com/onflow/cadence/issues/2541#issuecomment-1879211932. That one still need to be handled. I will open a follow-up PR.

Reply:
Added the second pat to the same PR, as it turns out to be a simple fix: https://github.com/onflow/cadence/pull/3132/commits/25b0726314c4cd02bd100ae7d41cffa36711f3d3